### ISSUE TITLE
Detect idiom: sign extension

### ISSUE THREAD

**Issue by KOLANICH** [avast/retdec#1022]
```c++
template<uint8_t n, uint8_t missing_bits>
NumericT<n + missing_bits> signExtend(NumericT<n> value){
	return (value << missingBits) >> missingBits;
}
```

**Comment by PeterMatula** [comment]
Again, as in #1019, a more detailed description of what is wrong and how it could/should be made better, together with input so it can be reproduced, would greatly help.

**Comment by KOLANICH** [comment]
>Is this some kind of compiler idiom?

Yes and no. Sign extension should be generated by a compiler when signed bit fields are used. It is also widely used manually, the provided exaple is only one of them, and in fact is implementation detail of compiler, it is unpirtable. Of course we need to detect them all. 

>Do you have some input on where it occurs?

One can compile the code like the mentioned. But it is not guaranteed to be sign extension. IDK if compilers even can detect this idiom and emit a more optimal code for it.

>What do we produce at the moment?

Sometimes bit operations, sometimes integer arithmetics with the same effect.

>What do you suggest we should produce? (Do I guess right that you propose a function call?)

Yeah. Add a function for each variant of the idiom detected.
